### PR TITLE
GODOT_single_root

### PR DESCRIPTION
**Pull request by aaronfranke**
Preview: https://github.com/aaronfranke/glTF/tree/GODOT_single_root/extensions/2.0/Vendor/GODOT_single_root

Many games engines import glTF scenes as objects with a single root node. Unity has prefabs with a single root GameObject, and Godot has scenes with a single root node. Since glTF allows defining multiple root nodes, engines will insert the glTF root nodes as children of the "real" root node, which makes it difficult for a glTF file to define the properties of the "real" root node. This is important for things like physics bodies and character controllers, which are defined on the root node, so that all child nodes are moved with the body. Aside from physics, a single root node would avoid an extra node in the tree when round-tripping between glTF and these engines.

The `GODOT_single_root` glTF extension solves this problem by imposing additional constraints on the glTF scene to ensure it can be parsed into one of these objects with a single root node. Implementations can detect `GODOT_single_root` and import the single root node as the object's "real" root node in the scene/prefab/etc. The extension contains no data, it only restricts behavior and gives a hint to importers.

If there is interest from other vendors, this could be made into a multi-vendor extension. The restrictions imposed here could also be added to the base glTF spec in a future version (could be, say, "2.1" since this can be done without breaking compatibility as long as we require the scene/scenes boilerplate).

EDIT: Also note, the restrictions imposed here are a superset of https://github.com/KhronosGroup/glTF/issues/1542.

### CODE REVIEW

File: extensions/2.0/Vendor/GODOT_single_root/GODOT_single_root.md
Comment (lexaknyazev) [inline extensions/2.0/Vendor/GODOT_single_root/GODOT_single_root.md:24]: These restrictions contradict the language above:
> import glTF scenes as objects with a single root node

and

> constraints on the glTF scene to ensure it can be parsed into one of these objects with a single root node

Either the overview needs to be adjusted to provide rationale for disallowing multiple scenes, which is orthogonal to root nodes, or these two restrictions should be lifted.

File: extensions/2.0/Vendor/GODOT_single_root/GODOT_single_root.md
Comment (lexaknyazev) [inline extensions/2.0/Vendor/GODOT_single_root/GODOT_single_root.md:26]: Order of elements in glTF object arrays has no specific meaning (except a GLB buffer reference for historical reasons). Extensions should not change that unless there's a very strong reason to.

File: extensions/2.0/Vendor/GODOT_single_root/GODOT_single_root.md
Comment (lexaknyazev) [inline extensions/2.0/Vendor/GODOT_single_root/GODOT_single_root.md:27]: These properties have default values so they may be always defined in some environments. Consider changing to:
> The ... properties of the single root node MUST either be undefined or have default values.

File: extensions/2.0/Vendor/GODOT_single_root/GODOT_single_root.md
Comment (lexaknyazev) [inline extensions/2.0/Vendor/GODOT_single_root/GODOT_single_root.md:19]: > which makes it difficult for a glTF file to define the properties of the "real" root node

What are these properties and why cannot scene objects be used for defining them?

File: extensions/2.0/Vendor/GODOT_single_root/GODOT_single_root.md
Comment (lexaknyazev) [inline extensions/2.0/Vendor/GODOT_single_root/GODOT_single_root.md:21]: The extension should mention whether it restricts nodes that do not belong to any scene.

**Review by lexaknyazev — CHANGES_REQUESTED** [review]

**Comment by javagl** [comment]
Just as a cross-reference: https://github.com/KhronosGroup/glTF/issues/1542

File: extensions/2.0/Vendor/GODOT_single_root/GODOT_single_root.md
Comment (aaronfranke) [reply extensions/2.0/Vendor/GODOT_single_root/GODOT_single_root.md:19]: Properties defined in glTF nodes, especially extensions. I don't think it makes sense to duplicate all of the node logic to apply to scenes, both import and export. You wouldn't add a `"mesh"` to a scene, right?

File: extensions/2.0/Vendor/GODOT_single_root/GODOT_single_root.md
Comment (aaronfranke) [reply extensions/2.0/Vendor/GODOT_single_root/GODOT_single_root.md:26]: The specific reason is to make the `"scene"` and `"scenes"` properties redundant, so the glTF file does not have any properties that specify which node is the root node. Since the goal is to have only root node, a list of root nodes is unnecessary and undesired, and is only required to keep for compatibility with glTF.

**Comment by aaronfranke** [comment]
Yes, the restrictions here are very much related to #1542. For further explanation of the problem and proposed solution, see my comment at the bottom of that issue: https://github.com/KhronosGroup/glTF/issues/1542#issuecomment-1675792594

**Review by emackey — APPROVED** [review]
Since this is a vendor extension used in production vendor software, we should probably merge this (and generally any other PRs in similar condition).  My approval here is not an endorsement for use of this in other software, just an acknowledgement that we should allow vendor-specific documentation of its use.

**Comment by aaronfranke** [comment]
What is the Khronos stance on merging vendor extensions? This is finalized and actively used in production.

**Comment by javagl** [comment]
Maybe an explicit ping @lexaknyazev can help. (I skimmed over the requested changes, and saw that they had been marked as 'resolved', but not by the _reviewer_ - so this PR will at least need another approval, to make sure that the changes actually are resolved _from the standpoint of the reviewer_). 

(No opinion from my side. The _goal_ of the extension itself sounds reasonable, as a somewhat standardized, non-breaking way of solving https://github.com/KhronosGroup/glTF/issues/1542 , and the goal of ~"making `scene/scenes` obsolete" sounds reasonable as well, but the devil is in the detail, and I trust Alexey to ensure that there is no breaking change or inconsistency hidden in the specification text)

**Comment by donmccurdy** [comment]
Hey all, I feel it's important that vendor extensions are approved and merged regularly (in the absence of egregious quality issues). If there's not bandwidth for a thorough review from working group members, that's entirely understandable! But perhaps the default in that case should be to merge extensions as-is, and perhaps even automatically. GitHub "labels" on extension PRs could be helpful as well, to indicate general PR status and next steps.

**Comment by emackey** [comment]
/ping @lexaknyazev This PR is actually blocked by your earlier request for changes. None of us have GitHub permission to merge this with that request still present here.  Can you remove the block?

**Comment by fire** [comment]
@lexaknyazev Can you reply and remove the block or at least state reasoning for keeping the block?

File: extensions/2.0/Vendor/GODOT_single_root/examples/dynamic_body_root.gltf
Comment (lexaknyazev) [inline extensions/2.0/Vendor/GODOT_single_root/examples/dynamic_body_root.gltf:1]: Update the example asset to use only this extension and add a link to it to the extension document OR just remove the example asset since the inline example present in the README is enough for showing the extension design.

**Comment by aaronfranke** [comment]
The example asset has been removed as requested.

**Comment by aaronfranke** [comment]
@lexaknyazev This PR is blocked from merging by your request for changes.

File: extensions/2.0/Vendor/GODOT_single_root/README.md
Comment (lexaknyazev) [inline extensions/2.0/Vendor/GODOT_single_root/README.md:33]: Use spaces for indentation.

File: extensions/2.0/Vendor/GODOT_single_root/README.md
Comment (lexaknyazev) [inline extensions/2.0/Vendor/GODOT_single_root/README.md:64]: Use spaces for indentation.

File: extensions/2.0/Vendor/GODOT_single_root/README.md
Comment (lexaknyazev) [inline extensions/2.0/Vendor/GODOT_single_root/README.md:38]: Remove this link; an extension should not make such comments about the base spec.

File: extensions/2.0/Vendor/GODOT_single_root/GODOT_single_root.md
Comment (lexaknyazev) [reply extensions/2.0/Vendor/GODOT_single_root/GODOT_single_root.md]: Nodes not belonging to any scene would violate the "single root" idea so the extension should be explicit about them.

Without addressing this question, the extension becomes completely redundant since explicitly doing
```js
if (gltf.scene === 0 &&
    gltf.scenes.length == 1 &&
    gltf.scenes[0].nodes.length == 1 &&
    gltf.scenes[0].nodes[0] === 0 &&
    gltf.nodes[0].hasDefaultTransform())
{
   singleRoot = true; // actually single root
}
```

is not more difficult but much more reliable than something like

```js
if (gltf.extensionsUsed.contains("GODOT_single_root "))
{
   singleRoot = true; // maybe single root
}
```

**Comment by aaronfranke** [comment]
@lexaknyazev Updated per your latest comments. Clarified that nodes are allowed to not be in the scene.

> Without addressing this question, the extension becomes completely redundant since explicitly doing

Those are all prerequisite conditions to being compliant with `GODOT_single_root`, but those checks alone would not be sufficient to have the desired behavior. Consider the following case:

```
glTF Scene
    glTF Node A
        glTF Node B
```

With `GODOT_single_root`, the imported root is "glTF Node A". Without, the root is "glTF Scene". We can't automatically make "glTF Node A" be the root for consistency with this case:

```
glTF Scene
    glTF Node A
        glTF Node B
    glTF Node C
```

In this case, we have no choice but to make "glTF Scene" be the root. We don't want the existence or non-existence of "glTF Node C" to affect the behavior of "glTF Node A" as the scene root or not. Therefore, we make an explicit flag for files to ask for "glTF Node A" to be the root, and this flag is only valid when the prerequisite conditions are true.

File: extensions/2.0/Vendor/GODOT_single_root/README.md
Comment (lexaknyazev) [inline extensions/2.0/Vendor/GODOT_single_root/README.md:21]: Does the updated language mean that out-of-scene nodes are silently ignored?

File: extensions/2.0/Vendor/GODOT_single_root/README.md
Comment (lexaknyazev) [inline extensions/2.0/Vendor/GODOT_single_root/README.md:38]: Please remove the "and is usually not a meaningful concept" part.

**Comment by lexaknyazev** [comment]
> Therefore, we make an explicit flag for files to ask for "glTF Node A" to be the root

So the goal is to explicitly flag assets that _should_ be imported without creating an extra "Scene" node assuming that they are compatible with such workflows. In other words, the main point of the extension is not to add some restrictions but to signal the intended asset usage that needs those restrictions.

Following the example above, the behaviour should then be:

| Node C | Single root flag | Action |
|---|---|---|
| ❌ | ❌ | Import scene as root |
| ✔️ | ❌ | Import scene as root |
| ❌ | ✔️ | Import Node A as root |
| ✔️ | ✔️ | Asset is invalid |

The first sentence of the extension overview should highlight this key detail (feel free to tweak the language):
> This extension allows flagging glTF assets as having a single root node to enable more efficient and accurate workflows with various DCC tools, e.g., by directly importing the asset's node hierarchy without creating additional scene nodes. This flag imposes certain constraints on glTF assets as described below.

File: extensions/2.0/Vendor/GODOT_single_root/README.md
Comment (aaronfranke) [reply extensions/2.0/Vendor/GODOT_single_root/README.md:21]: Aren't they already silently ignored with base glTF, unless some other extension uses them?

Comment (javagl) [reply extensions/2.0/Vendor/GODOT_single_root/README.md]: ^ That _might_ have referred to _skeletons_, but I haven't read the details, so that might be completely off-track. 

Iff it referred to skeletons, then maybe 

> all other nodes are either descendants of this node

could be

> all other _non-skeleton_ nodes are either descendants of this node

or so, but all this is to be confirmed - maybe I mistunderstood something here

Comment (aaronfranke) [reply extensions/2.0/Vendor/GODOT_single_root/README.md]: I don't want to impose a restriction either which way for non-scene nodes. They should behave the same as a vanilla glTF asset.

Comment (lexaknyazev) [reply extensions/2.0/Vendor/GODOT_single_root/README.md]: Ack, assuming that non-scene node presence does not lead to creating a synthetic "Scene-for-non-Scene-nodes" root.

**Review by lexaknyazev — APPROVED** [review]
LGTM since the extension intention is clearly stated now.

**Pull request merged**
